File: problems/tier2-intermediate/017-lru-cache/solution.py
"""LRU Cache with TTL expiry and eviction listeners."""

from collections import OrderedDict


class LruOp:
    def __init__(self, kind, i1=0, i2=0, i3=0, i4=0):
        self.kind = kind
        self.i1 = i1
        self.i2 = i2
        self.i3 = i3
        self.i4 = i4


# Eviction reasons
CAPACITY = "CAPACITY"
TTL_EXPIRED = "TTL_EXPIRED"
EXPLICIT_DELETE = "EXPLICIT_DELETE"


class CapturingListener:
    def __init__(self):
        self.events = []  # list of (key, value, reason)

    def onEviction(self, key, value, reason):
        self.events.append((key, value, reason))


class LRUCache:
    def __init__(self, capacity):
        self.capacity = capacity
        # OrderedDict: key -> (value, expires_at)  (move_to_end -> most recent)
        # We treat the FIRST item as LRU, LAST as MRU.
        self.data = OrderedDict()
        self.listeners = []

    def _is_expired(self, expires_at, current_time):
        return expires_at > 0 and current_time >= expires_at

    def _notify(self, key, value, reason):
        for listener in self.listeners:
            listener.onEviction(key, value, reason)

    def _evict_key(self, key, reason):
        value, _ = self.data[key]
        del self.data[key]
        self._notify(key, value, reason)

    def _evict_lru(self):
        if not self.data:
            return
        key = next(iter(self.data))
        self._evict_key(key, CAPACITY)

    # Part 1
    def get(self, key):
        if key not in self.data:
            return -1
        self.data.move_to_end(key)
        return self.data[key][0]

    def put(self, key, value):
        if key in self.data:
            _, exp = self.data[key]
            self.data[key] = (value, exp)
            self.data.move_to_end(key)
            return
        if len(self.data) >= self.capacity:
            self._evict_lru()
        self.data[key] = (value, 0)

    # Part 2 (TTL)
    def get_t(self, key, current_time):
        if key not in self.data:
            return -1
        value, exp = self.data[key]
        if self._is_expired(exp, current_time):
            self._evict_key(key, TTL_EXPIRED)
            return -1
        self.data.move_to_end(key)
        return value

    def put_t(self, key, value, current_time, ttl=0):
        expires_at = current_time + ttl if ttl > 0 else 0
        if key in self.data:
            _, exp = self.data[key]
            if self._is_expired(exp, current_time):
                self._evict_key(key, TTL_EXPIRED)
            else:
                self.data[key] = (value, expires_at)
                self.data.move_to_end(key)
                return
        if len(self.data) >= self.capacity:
            self._evict_lru()
        self.data[key] = (value, expires_at)

    def delete_key(self, key):
        if key not in self.data:
            return False
        self._evict_key(key, EXPLICIT_DELETE)
        return True

    def size(self):
        return len(self.data)

    # Part 3
    def add_eviction_listener(self, listener):
        self.listeners.append(listener)

    def remove_eviction_listener(self, listener):
        self.listeners = [l for l in self.listeners if l is not listener]


def lru_simulate(ops):
    out = []
    cache = None
    listeners = []

    def ensure_listener(idx):
        while len(listeners) <= idx:
            listeners.append(CapturingListener())

    for op in ops:
        k = op.kind
        if k == "new":
            cache = LRUCache(op.i1)
            listeners = []
            out.append("ok")
        elif k == "put":
            cache.put(op.i1, op.i2)
            out.append("ok")
        elif k == "put_t":
            cache.put_t(op.i1, op.i2, op.i3, op.i4)
            out.append("ok")
        elif k == "get":
            out.append(str(cache.get(op.i1)))
        elif k == "get_t":
            out.append(str(cache.get_t(op.i1, op.i2)))
        elif k == "delete":
            out.append("ok" if cache.delete_key(op.i1) else "fail")
        elif k == "size":
            out.append(str(cache.size()))
        elif k == "add_listener":
            ensure_listener(op.i1)
            cache.add_eviction_listener(listeners[op.i1])
            out.append("ok")
        elif k == "remove_listener":
            if op.i1 < len(listeners):
                cache.remove_eviction_listener(listeners[op.i1])
            out.append("ok")
        elif k == "events_count":
            if op.i1 < len(listeners):
                out.append(str(len(listeners[op.i1].events)))
            else:
                out.append("0")
        elif k == "event_key":
            if op.i1 < len(listeners) and op.i2 < len(listeners[op.i1].events):
                out.append(str(listeners[op.i1].events[op.i2][0]))
            else:
                out.append("")
        elif k == "event_value":
            if op.i1 < len(listeners) and op.i2 < len(listeners[op.i1].events):
                out.append(str(listeners[op.i1].events[op.i2][1]))
            else:
                out.append("")
        elif k == "event_reason":
            if op.i1 < len(listeners) and op.i2 < len(listeners[op.i1].events):
                out.append(listeners[op.i1].events[op.i2][2])
            else:
                out.append("")
        else:
            out.append("unknown:" + k)
    return out

File: problems/tier2-intermediate/017-lru-cache/solution.js
"use strict";
// LRU Cache with TTL expiry and eviction listeners.

class LruOp {
  constructor(kind, i1 = 0, i2 = 0, i3 = 0, i4 = 0) {
    this.kind = kind;
    this.i1 = i1;
    this.i2 = i2;
    this.i3 = i3;
    this.i4 = i4;
  }
}

// Eviction reasons
const CAPACITY = "CAPACITY";
const TTL_EXPIRED = "TTL_EXPIRED";
const EXPLICIT_DELETE = "EXPLICIT_DELETE";

class CapturingListener {
  constructor() {
    this.events = []; // list of [key, value, reason]
  }
  onEviction(key, value, reason) {
    this.events.push([key, value, reason]);
  }
}

class LRUCache {
  constructor(capacity) {
    this.capacity = capacity;
    // Map: key -> [value, expires_at]. Insertion order: FIRST is LRU, LAST is MRU.
    this.data = new Map();
    this.listeners = [];
  }

  _isExpired(expiresAt, currentTime) {
    return expiresAt > 0 && currentTime >= expiresAt;
  }

  _notify(key, value, reason) {
    for (const listener of this.listeners) {
      listener.onEviction(key, value, reason);
    }
  }

  _evictKey(key, reason) {
    const [value] = this.data.get(key);
    this.data.delete(key);
    this._notify(key, value, reason);
  }

  _evictLru() {
    if (this.data.size === 0) return;
    const key = this.data.keys().next().value;
    this._evictKey(key, CAPACITY);
  }

  // Part 1
  get(key) {
    if (!this.data.has(key)) return -1;
    const entry = this.data.get(key);
    this.data.delete(key);
    this.data.set(key, entry);
    return entry[0];
  }

  put(key, value) {
    if (this.data.has(key)) {
      const [, exp] = this.data.get(key);
      this.data.delete(key);
      this.data.set(key, [value, exp]);
      return;
    }
    if (this.data.size >= this.capacity) {
      this._evictLru();
    }
    this.data.set(key, [value, 0]);
  }

  // Part 2 (TTL)
  get_t(key, currentTime) {
    if (!this.data.has(key)) return -1;
    const [value, exp] = this.data.get(key);
    if (this._isExpired(exp, currentTime)) {
      this._evictKey(key, TTL_EXPIRED);
      return -1;
    }
    const entry = this.data.get(key);
    this.data.delete(key);
    this.data.set(key, entry);
    return value;
  }

  put_t(key, value, currentTime, ttl = 0) {
    const expiresAt = ttl > 0 ? currentTime + ttl : 0;
    if (this.data.has(key)) {
      const [, exp] = this.data.get(key);
      if (this._isExpired(exp, currentTime)) {
        this._evictKey(key, TTL_EXPIRED);
      } else {
        this.data.delete(key);
        this.data.set(key, [value, expiresAt]);
        return;
      }
    }
    if (this.data.size >= this.capacity) {
      this._evictLru();
    }
    this.data.set(key, [value, expiresAt]);
  }

  delete_key(key) {
    if (!this.data.has(key)) return false;
    this._evictKey(key, EXPLICIT_DELETE);
    return true;
  }

  size() {
    return this.data.size;
  }

  // Part 3
  add_eviction_listener(listener) {
    this.listeners.push(listener);
  }

  remove_eviction_listener(listener) {
    this.listeners = this.listeners.filter((l) => l !== listener);
  }
}

function lru_simulate(ops) {
  const out = [];
  let cache = null;
  let listeners = [];

  const ensureListener = (idx) => {
    while (listeners.length <= idx) {
      listeners.push(new CapturingListener());
    }
  };

  for (const op of ops) {
    const k = op.kind;
    if (k === "new") {
      cache = new LRUCache(op.i1);
      listeners = [];
      out.push("ok");
    } else if (k === "put") {
      cache.put(op.i1, op.i2);
      out.push("ok");
    } else if (k === "put_t") {
      cache.put_t(op.i1, op.i2, op.i3, op.i4);
      out.push("ok");
    } else if (k === "get") {
      out.push(String(cache.get(op.i1)));
    } else if (k === "get_t") {
      out.push(String(cache.get_t(op.i1, op.i2)));
    } else if (k === "delete") {
      out.push(cache.delete_key(op.i1) ? "ok" : "fail");
    } else if (k === "size") {
      out.push(String(cache.size()));
    } else if (k === "add_listener") {
      ensureListener(op.i1);
      cache.add_eviction_listener(listeners[op.i1]);
      out.push("ok");
    } else if (k === "remove_listener") {
      if (op.i1 < listeners.length) {
        cache.remove_eviction_listener(listeners[op.i1]);
      }
      out.push("ok");
    } else if (k === "events_count") {
      if (op.i1 < listeners.length) {
        out.push(String(listeners[op.i1].events.length));
      } else {
        out.push("0");
      }
    } else if (k === "event_key") {
      if (op.i1 < listeners.length && op.i2 < listeners[op.i1].events.length) {
        out.push(String(listeners[op.i1].events[op.i2][0]));
      } else {
        out.push("");
      }
    } else if (k === "event_value") {
      if (op.i1 < listeners.length && op.i2 < listeners[op.i1].events.length) {
        out.push(String(listeners[op.i1].events[op.i2][1]));
      } else {
        out.push("");
      }
    } else if (k === "event_reason") {
      if (op.i1 < listeners.length && op.i2 < listeners[op.i1].events.length) {
        out.push(listeners[op.i1].events[op.i2][2]);
      } else {
        out.push("");
      }
    } else {
      out.push("unknown:" + k);
    }
  }
  return out;
}

module.exports = { LruOp, lru_simulate };
